File: bench/julia/bench.jl
# Cross-language micro-benchmark harness (Julia / MultiFloats.jl).
#
# The input generation and the timing methodology are shared with
# `benches/core.rs` and `bench/python/bench.py`; the canonical specification
# lives in `bench/README.md`.  Deliberately dependency-free apart from
# MultiFloats: one CSV row per `(operation, mode)`.
#
# Usage:
#
#     taskset -c 2 julia --project=bench/julia bench/julia/bench.jl \
#         --out bench/out/julia.csv
#
# Copyright (C) 2023-2025 Markus Wallerberger and others
# SPDX-License-Identifier: MIT

using MultiFloats
using Printf

# ---------------------------------------------------------------------------
# Shared input specification

const SPLITMIX_INC = 0x9E3779B97F4A7C15
const FNV_OFFSET = 0xcbf29ce484222325
const FNV_PRIME = 0x00000100000001b3
const POWI_EXP = 3

function splitmix64(state::UInt64)
    state += SPLITMIX_INC
    z = state
    z = (z ⊻ (z >> 30)) * 0xBF58476D1CE4E5B9
    z = (z ⊻ (z >> 27)) * 0x94D049BB133111EB
    return state, z ⊻ (z >> 31)
end

function fnv1a64(s::AbstractString)
    h = FNV_OFFSET
    for b in codeunits(s)
        h ⊻= UInt64(b)
        h *= FNV_PRIME
    end
    return h
end

# Uniform value in [1, 2) built from the 52 low bits of `u`.
@inline unit12(u::UInt64) = reinterpret(Float64, 0x3FF0000000000000 | (u >> 12))

# Uniform value in [-0.5, 0.5).
@inline signed12(u::UInt64) = unit12(u) - 1.5

# Exact power of two 2^-k for k in [0, 21).
@inline pow2_neg(k::Int) = 1.0 / Float64(UInt64(1) << k)

# ---------------------------------------------------------------------------
# Operations

const OPS = ["noop", "add", "sub", "mul", "div", "sqrt", "cbrt",
             "exp", "exp2", "log", "log2", "log10", "powi", "powf"]

const BINARY = Set(["add", "sub", "mul", "div", "powf"])

# The input transforms cover every operation of the shared specification, not
# only the ones this harness measures, so that the three generators stay
# literally identical and the checksum comparison keeps its meaning when an
# operation is added here later.
function input_a(op::AbstractString, u::UInt64)
    if op == "exp" || op == "exp2"
        return signed12(u) * 32.0
    elseif op in ("sin", "cos", "tan", "atan", "atan2")
        return signed12(u) * 8.0
    elseif op in ("sinh", "cosh", "tanh")
        return signed12(u) * 4.0
    elseif op == "expm1" || op == "log1p"
        return signed12(u) * 0.0625
    else
        return unit12(u)
    end
end

function input_b(op::AbstractString, i::Int, u::UInt64)
    if op == "add" || op == "sub"
        return unit12(u) * pow2_neg((i - 1) % 21)
    elseif op == "atan2"
        return signed12(u) * 8.0
    elseif op == "powf"
        return signed12(u) * 4.0
    else
        return unit12(u)
    end
end

@inline function _hash_update(hash::UInt64, x::Float64)
    bits = reinterpret(UInt64, x)
    for k in 0:7
        hash ⊻= (bits >> (8 * k)) & UInt64(0xFF)
        hash *= FNV_PRIME
    end
    return hash
end

function generate(op::AbstractString, n::Int)
    state = fnv1a64(op)
    hash = FNV_OFFSET

    a = Vector{Float64}(undef, n)
    for i in 1:n
        state, u = splitmix64(state)
        a[i] = input_a(op, u)
        hash = _hash_update(hash, a[i])
    end

    b = Vector{Float64}(undef, n)
    for i in 1:n
        state, u = splitmix64(state)
        b[i] = input_b(op, i, u)
        hash = _hash_update(hash, b[i])
    end

    return a, b, hash
end

# ---------------------------------------------------------------------------
# Evaluation
#
# The dispatch on the operation name happens once, outside the timed loop
# (function barrier); the returned closure is specialised by the compiler.

function op_function(op::AbstractString)
    if op == "noop"
        (a, b) -> a
    elseif op == "add"
        (a, b) -> a + b
    elseif op == "sub"
        (a, b) -> a - b
    elseif op == "mul"
        (a, b) -> a * b
    elseif op == "div"
        (a, b) -> a / b
    elseif op == "sqrt"
        (a, b) -> sqrt(a)
    elseif op == "cbrt"
        (a, b) -> cbrt(a)
    elseif op == "exp"
        (a, b) -> exp(a)
    elseif op == "exp2"
        (a, b) -> exp2(a)
    elseif op == "log"
        (a, b) -> log(a)
    elseif op == "log2"
        (a, b) -> log2(a)
    elseif op == "log10"
        (a, b) -> log10(a)
    elseif op == "powi"
        (a, b) -> a^POWI_EXP
    elseif op == "powf"
        (a, b) -> a^b
    else
        error("unknown op $op")
    end
end

# Reduce a benchmark value to a plain Float64 so that the accumulation loop is
# identical for both element types.  `_limbs[1]` is the leading limb.
@inline reduce_value(x::Float64) = x
@inline reduce_value(x::Float64x2) = x._limbs[1]

# ---------------------------------------------------------------------------
# Timing

const SINKS = Ref(0.0)

function median_sample!(samples::Vector{Float64})
    sort!(samples)
    return samples[cld(length(samples), 2)]
end

"""Batched (throughput) form: independent operations, four-way unrolled."""
function time_throughput(n::Int, reps::Int, eval::F) where {F}
    samples = Vector{Float64}(undef, reps)
    for r in 1:reps
        s0 = 0.0
        s1 = 0.0
        s2 = 0.0
        s3 = 0.0
        t0 = time_ns()
        i = 1
        while i + 3 <= n
            s0 += reduce_value(eval(i))
            s1 += reduce_value(eval(i + 1))
            s2 += reduce_value(eval(i + 2))
            s3 += reduce_value(eval(i + 3))
            i += 4
        end
        while i <= n
            s0 += reduce_value(eval(i))
            i += 1
        end
        dt = (time_ns() - t0) / n
        SINKS[] += s0 + s1 + s2 + s3
        samples[r] = dt
    end
    return median_sample!(samples)
end

"""Dependent (latency) form.  Informational only."""
function time_latency(op::AbstractString, a::Vector{T}, b::Vector{T}, reps::Int) where {T}
    n = length(a)
    binary = op in BINARY
    f = op_function(op)
    samples = Vector{Float64}(undef, reps)
    for r in 1:reps
        acc = one(T)
        t0 = time_ns()
        if binary
            for i in 1:n
                acc = f(a[i], acc)
            end
        else
            for i in 1:n
                acc = f(acc, b[i])
            end
        end
        dt = (time_ns() - t0) / n
        SINKS[] += reduce_value(acc)
        samples[r] = dt
    end
    return median_sample!(samples)
end

# ---------------------------------------------------------------------------
# Driver

function main()
    out = "bench/out/julia.csv"
    n = 16384
    reps = 15
    argv = ARGS
    i = 1
    while i <= length(argv)
        if argv[i] == "--out"
            out = argv[i+1]; i += 2
        elseif argv[i] == "--n"
            n = parse(Int, argv[i+1]); i += 2
        elseif argv[i] == "--reps"
            reps = parse(Int, argv[i+1]); i += 2
        else
            error("unknown argument: $(argv[i])")
        end
    end

    mkpath(dirname(out))
    rows = String["impl,op,mode,ns_per_op,checksum"]

    println("impl=multifloats n=$n reps=$reps")
    @printf("%-8s %12s %12s %18s\n", "op", "mf-thr", "mf-lat", "checksum")

    for op in OPS
        a64, b64, checksum = generate(op, n)
        amf = Float64x2.(a64)
        bmf = Float64x2.(b64)

        f = op_function(op)
        thr = time_throughput(n, reps, i -> f(amf[i], bmf[i]))
        lat = time_latency(op, amf, bmf, reps)

        @printf("%-8s %12.4f %12.4f 0x%016x\n", op, thr, lat, checksum)
        cs = @sprintf("0x%016x", checksum)
        push!(rows, "multifloats,$op,throughput,$(round(thr, digits=6)),$cs")
        push!(rows, "multifloats,$op,latency,$(round(lat, digits=6)),$cs")
    end

    open(out, "w") do io
        println(io, "# xprec-rs benchmark harness (julia/MultiFloats)")
        println(io, "# impl=multifloats n=$n reps=$reps powi_exp=$POWI_EXP")
        for row in rows
            println(io, row)
        end
    end
    println("\nwrote $out")
end

main()

File: bench/python/bench.py
#!/usr/bin/env python3
"""Cross-language micro-benchmark harness (Python / numpy + xprec.ddouble).

The input generation and the timing methodology are shared with
``benches/core.rs`` and ``bench/julia/bench.jl``; the canonical specification
lives in ``bench/README.md``.

Only operations that the `xprec` numpy extension actually provides as ufuncs
are measured; `exp2`, `log2`, `log10` and `cbrt` have no `ddouble` ufunc and
are therefore reported as not available.

Usage:

    taskset -c 2 python bench/python/bench.py --out bench/out/python.csv

Copyright (C) 2023-2025 Markus Wallerberger and others
SPDX-License-Identifier: MIT
"""

import argparse
import os
import time

import numpy as np

import xprec

MASK = 0xFFFFFFFFFFFFFFFF
SPLITMIX_INC = 0x9E3779B97F4A7C15
FNV_OFFSET = 0xCBF29CE484222325
FNV_PRIME = 0x00000100000001B3
POWI_EXP = 3

DD = xprec.ddouble
SINK = 0.0


# ---------------------------------------------------------------------------
# Shared input specification


def splitmix64(state):
    """Scalar splitmix64; used only for the per-operation seed."""
    state = (state + SPLITMIX_INC) & MASK
    z = state
    z = ((z ^ (z >> 30)) * 0xBF58476D1CE4E5B9) & MASK
    z = ((z ^ (z >> 27)) * 0x94D049BB133111EB) & MASK
    return state, z ^ (z >> 31)


def fnv1a64(text):
    h = FNV_OFFSET
    for byte in text.encode():
        h ^= byte
        h = (h * FNV_PRIME) & MASK
    return h


def _splitmix64_stream(seed, count):
    """Vectorised splitmix64 stream, identical to the scalar sequence."""
    k = np.arange(1, count + 1, dtype=np.uint64)
    state = np.uint64(seed) + k * np.uint64(SPLITMIX_INC)
    z = state
    z = (z ^ (z >> np.uint64(30))) * np.uint64(0xBF58476D1CE4E5B9)
    z = (z ^ (z >> np.uint64(27))) * np.uint64(0x94D049BB133111EB)
    return z ^ (z >> np.uint64(31))


def unit12(z):
    """Uniform value in [1, 2) built from the 52 low bits of `z`."""
    bits = np.uint64(0x3FF0000000000000) | (z >> np.uint64(12))
    return bits.view(np.float64)


def signed12(z):
    """Uniform value in [-0.5, 0.5)."""
    return unit12(z) - 1.5


def pow2_neg(k):
    """Exact power of two 2^-k for k in [0, 21)."""
    return 1.0 / (np.uint64(1) << k.astype(np.uint64)).astype(np.float64)


def input_a(op, z):
    if op in ("exp", "exp2"):
        return signed12(z) * 32.0
    if op in ("sin", "cos", "tan", "atan", "atan2"):
        return signed12(z) * 8.0
    if op in ("sinh", "cosh", "tanh"):
        return signed12(z) * 4.0
    if op in ("expm1", "log1p"):
        return signed12(z) * 0.0625
    return unit12(z)


def input_b(op, z, i):
    if op in ("add", "sub"):
        return unit12(z) * pow2_neg(i % 21)
    if op == "atan2":
        return signed12(z) * 8.0
    if op == "powf":
        return signed12(z) * 4.0
    return unit12(z)


def generate(op, n):
    seed = fnv1a64(op)
    z = _splitmix64_stream(seed, 2 * n)
    za, zb = z[:n], z[n:]
    idx = np.arange(n)
    a = input_a(op, za)
    b = input_b(op, zb, idx)
    return a, b, checksum(a, b)


def checksum(a, b):
    """FNV-1a over the little-endian bytes of all generated values."""
    h = FNV_OFFSET
    for byte in np.concatenate([a, b]).tobytes():
        h ^= byte
        h = (h * FNV_PRIME) & MASK
    return h


# ---------------------------------------------------------------------------
# Operations
#
# Each entry is `(n_arguments, callable)`.  `powi` uses the same fixed
# integer exponent as the other harnesses; note that numpy applies the
# general `power` ufunc here, so the Python number is not a dedicated
# integer-power kernel.

OPS = {
    "add": (2, lambda x, y: x + y),
    "sub": (2, lambda x, y: x - y),
    "mul": (2, lambda x, y: x * y),
    "div": (2, lambda x, y: x / y),
    "sqrt": (1, np.sqrt),
    "exp": (1, np.exp),
    "log": (1, np.log),
    "powi": (1, lambda x: np.power(x, POWI_EXP)),
    "powf": (2, np.power),
    "sin": (1, np.sin),
    "cos": (1, np.cos),
    "tan": (1, np.tan),
    "atan": (1, np.arctan),
    "atan2": (2, np.arctan2),
    "sinh": (1, np.sinh),
    "cosh": (1, np.cosh),
    "tanh": (1, np.tanh),
    "expm1": (1, np.expm1),
}


# ---------------------------------------------------------------------------
# Timing


def median_sample(samples):
    samples.sort()
    return samples[len(samples) // 2]


def time_batched(f, x, y, n, reps):
    """Batched form: one whole-array ufunc call, timed with `perf_counter_ns`."""
    global SINK
    f(x, y)  # warm up
    samples = []
    for _ in range(reps):
        t0 = time.perf_counter_ns()
        out = f(x, y)
        t1 = time.perf_counter_ns()
        SINK += float(out.reshape(-1)[0])
        samples.append((t1 - t0) / n)
    return median_sample(samples)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--out", default="bench/out/python.csv")
    parser.add_argument("--n", type=int, default=16384)
    parser.add_argument("--reps", type=int, default=15)
    args = parser.parse_args()

    os.makedirs(os.path.dirname(args.out) or ".", exist_ok=True)
    rows = ["impl,op,mode,ns_per_op,checksum"]

    print(f"impl=numpy-xprec n={args.n} reps={args.reps}")
    print(f"{'op':<8} {'py-thr':>12} {'checksum':>18}")

    for op, (nargs, f) in OPS.items():
        a64, b64, cs = generate(op, args.n)
        x = a64.astype(DD)
        y = b64.astype(DD)

        if nargs == 1:
            thr = time_batched(lambda p, q: f(p), x, y, args.n, args.reps)
        else:
            thr = time_batched(f, x, y, args.n, args.reps)

        print(f"{op:<8} {thr:12.4f} {cs:#018x}")
        rows.append(f"numpy-xprec,{op},throughput,{thr:.6f},{cs:#018x}")

    with open(args.out, "w", encoding="utf-8") as handle:
        handle.write("# xprec-rs benchmark harness (python/numpy+xprec)\n")
        handle.write(f"# impl=numpy-xprec n={args.n} reps={args.reps} powi_exp={POWI_EXP}\n")
        for row in rows:
            handle.write(row + "\n")
    print(f"\nwrote {args.out}")
    print(f"sink={SINK}")


if __name__ == "__main__":
    main()
